reference/ast-examples.md

```diff
@@ -0,0 +1,185 @@
+
+* [Explain how ASTs work at a high-level](#high-level-ast-explanation)
+* [Provide a sample chunk of code to demonstrate ASTs](#code)
+* Discuss how that code relates to different types of ASTs
+  * [Binary](#binary)
+  * [Block](#block)
+  * [ClassDeclaration](#classdeclaration)
+  * [CompilationUnit](#compilationunit)
+  * [Expression](#expression)
+  * [FieldAccess](#fieldaccess)
+  * [Identifier](#identifier)
+  * [MethodDeclaration](#methoddeclaration)
+  * [MethodInvocation](#methodinvocation)
+  * [NewClass](#newclass)
+  * [Statement](#statement)
+  * [VariableDeclarations](#variabledeclarations)
+* [Teach you how to use the debugger to learn more about ASTs yourself](#using-the-debugger-to-detect-asts)
+
+## High-level AST Explanation
+
+In order for a computer to parse and change code, there needs to be some data structure defined that can accurately represent said code. OpenRewrite uses an [Abstract Syntax Tree](/v1beta/abstract-syntax-trees.md) (AST) for this purpose. Like many tree data structures, ASTs are often composed of many other ASTs. 
+
+For instance, you could have a [CompilationUnit](#compilationunit) which then has a [ClassDeclaration](#classdeclaration) inside of that which then has a [MethodDeclaration](#methoddeclaration) inside of that and so on. On top of being a valid AST that would _also_ represent valid Java code. 
+
+On the other hand, you _could_ define a [Binary](#binary) with other Binaries inside of that. That would be a valid AST but it **would not** represent valid Java code as just entering `1+2` (a valid Binary) in a Java file would not compile (that code would need to be in an expression which is then in a class to be valid).
+
+> Every object in the [code section](#code) is an AST in the sense that it implements a Tree interface. These objects, however, need to be combined in an appropriate way to represent valid Java. It is your duty as a recipe author to ensure that the final AST represents valid Java code.
+
+## Code
+
+Below is a simple Java class whose entire purpose is to demonstrate different types of ASTs. Each of the following sections will highlight different parts of this code to demonstrate which chunks correspond to which AST.
+
+```java
+package org.openrewrite;
+                
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@AnAnnotation
+public class A {
+
+    List<Integer> a = new ArrayList<>();
+
+    int foo() {
+        int a = 1 + 2, b = 3, c;
+        this.a = this.a.stream()
+                .map(it -> it + 1)
+                .collect(Collectors.toList());
+        return a;
+    }
+}
+
+@interface AnAnnotation {}
+```
+
+### Binary
+
+A [Binary AST](https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/tree/J.java#L597-L705) is an [Expression](#expression) with a left and right side, separated by an operator. Examples of operators include `+`, `-`, `||`, `&&`, and more.
```

Comment:
The line numbers in these links will be wrong as soon as we make changes to `J`. Which aren't constant, but it does happen. Just today I tweaked an entry in `J`. If we can't make more durable links, and I'm not aware of any way to do that, better to link to `J` once without trying to link to its individual components.

Reply:
I could update the links to use a tag such as: https://github.com/openrewrite/rewrite/blob/v7.33.0/rewrite-java/src/main/java/org/openrewrite/java/tree/J.java which will stay consistent regardless of the changes made. 

Does that seem reasonable to you or would you prefer I just link to the file rather than the individual parts?